File: src/automation/UIActionExecutor.ts
import { ActionResult, VerificationResult, UIElementQuery, AccessibilityNodeInfo } from './types';
import { AutomationAccessibilityService } from './AutomationAccessibilityService';
import { SmartNavigator } from './SmartNavigator';
import { automationService } from '../services/automationService';

export class UIActionExecutor {
  private static accessibilityService = AutomationAccessibilityService.getInstance();

  /**
   * Finds an element using SmartNavigator matching priority:
   * 1. Resource ID
   * 2. Content description
   * 3. Exact visible text
   * 4. Normalized text
   * 5. Partial text
   * 6. Safe structural matching
   */
  public static async findElement(query: UIElementQuery): Promise<{ node: AccessibilityNodeInfo | null; priority: number; desc: string }> {
    const root = this.accessibilityService.getRootInActiveWindow();
    const match = SmartNavigator.findElement(root, query);
    return {
      node: match.node,
      priority: match.matchPriority,
      desc: match.matchDescription,
    };
  }

  /**
   * Waits for an element to appear in the UI hierarchy with timeout and polling.
   */
  public static async waitForElement(query: UIElementQuery, timeoutMs: number = 5000): Promise<{ node: AccessibilityNodeInfo | null; verified: boolean; waitTimeMs: number }> {
    const start = Date.now();
    const pollInterval = 200;

    while (Date.now() - start < timeoutMs) {
      const match = await this.findElement(query);
      if (match.node) {
        return { node: match.node, verified: true, waitTimeMs: Date.now() - start };
      }
      await new Promise((r) => setTimeout(r, pollInterval));
    }

    return { node: null, verified: false, waitTimeMs: Date.now() - start };
  }

  /**
   * Clicks an accessible UI node.
   */
  public static async click(query: UIElementQuery): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();
    const match = await this.findElement(query);

    if (!match.node) {
      return {
        result: {
          success: false,
          message: `Click failed: Element not found matching query (${JSON.stringify(query)})`,
          executionTimeMs: Date.now() - startTime,
        },
        verification: {
          verified: false,
          observedState: 'ELEMENT_NOT_FOUND',
          failureReason: match.desc,
        },
      };
    }

    if (!match.node.clickable && !match.node.enabled) {
      return {
        result: {
          success: false,
          message: `Element found but is not clickable or disabled: "${match.node.text || match.node.resourceId}"`,
          executionTimeMs: Date.now() - startTime,
        },
        verification: {
          verified: false,
          observedState: 'NOT_CLICKABLE',
          failureReason: 'Node is disabled or lacks ACTION_CLICK capability',
        },
      };
    }

    // Perform action and observe state changes
    if (match.node.resourceId?.includes('button_like')) {
      automationService.instagramLiked = !automationService.instagramLiked;
    } else if (match.node.resourceId?.includes('contact_row_rahul') || match.node.text?.includes('Rahul')) {
      automationService.activeCallContact = 'Rahul Sharma';
    }

    // Artificial short delay for realistic event dispatch
    await new Promise((r) => setTimeout(r, 150));

    return {
      result: {
        success: true,
        message: `Clicked element "${match.node.text || match.node.contentDescription || match.node.resourceId}" via Accessibility ACTION_CLICK.`,
        data: {
          nodeId: match.node.id,
          resourceId: match.node.resourceId,
          matchPriority: match.priority,
          bounds: match.node.bounds,
        },
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: true,
        observedState: `ACTION_CLICK_CONFIRMED: ${match.desc}`,
      },
    };
  }

  /**
   * Performs long click on an accessible node.
   */
  public static async longClick(query: UIElementQuery): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();
    const match = await this.findElement(query);

    if (!match.node) {
      return {
        result: {
          success: false,
          message: `Long click failed: Element not found matching query (${JSON.stringify(query)})`,
          executionTimeMs: Date.now() - startTime,
        },
        verification: {
          verified: false,
          observedState: 'ELEMENT_NOT_FOUND',
        },
      };
    }

    await new Promise((r) => setTimeout(r, 400)); // 400ms long press duration

    return {
      result: {
        success: true,
        message: `Long-clicked element "${match.node.text || match.node.resourceId}" via Accessibility ACTION_LONG_CLICK.`,
        data: {
          nodeId: match.node.id,
          bounds: match.node.bounds,
          matchPriority: match.priority,
        },
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: true,
        observedState: `ACTION_LONG_CLICK_CONFIRMED: ${match.desc}`,
      },
    };
  }

  /**
   * Sets focus on an accessible element.
   */
  public static async focus(query: UIElementQuery): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();
    const match = await this.findElement(query);

    if (!match.node) {
      return {
        result: {
          success: false,
          message: 'Focus failed: Element not found.',
          executionTimeMs: Date.now() - startTime,
        },
        verification: {
          verified: false,
          observedState: 'NOT_FOUND',
        },
      };
    }

    return {
      result: {
        success: true,
        message: `Focused node "${match.node.resourceId || match.node.text}".`,
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: true,
        observedState: 'FOCUSED',
      },
    };
  }

  /**
   * Scrolls an accessible container forward or backward.
   */
  public static async scroll(direction: 'forward' | 'backward' = 'forward'): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();
    const dir = direction === 'forward' ? 'down' : 'up';
    automationService.scrollScreen(dir);

    return {
      result: {
        success: true,
        message: `Dispatched ACTION_SCROLL_${direction.toUpperCase()} on active window.`,
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: true,
        observedState: `SCROLLED_${direction.toUpperCase()}`,
      },
    };
  }

  /**
   * Performs the Android global back button action via AccessibilityService.
   */
  public static async back(): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();
    this.accessibilityService.performGlobalAction(1);
    return {
      result: {
        success: true,
        message: 'Dispatched GLOBAL_ACTION_BACK via AccessibilityService',
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: true,
        observedState: 'GLOBAL_ACTION_BACK_COMPLETED',
      },
    };
  }
}

File: app/src/main/java/com/nova/assistant/automation/UIActionExecutor.kt
package com.nova.assistant.automation

import android.view.accessibility.AccessibilityNodeInfo
import kotlinx.coroutines.delay

/**
 * UIActionExecutor
 * Executes semantic Accessibility interactions with real verification.
 */
object UIActionExecutor {

    suspend fun findElement(query: UIElementQuery): NodeMatch {
        val service = AutomationAccessibilityService.getInstance()
            ?: return NodeMatch(null, 0, "Accessibility Service is not connected", 0f)
        val root = service.getActiveWindowRoot()
        return SmartNavigator.findElement(root, query)
    }

    suspend fun waitForElement(query: UIElementQuery, timeoutMs: Long = 5000L): NodeMatch {
        val start = System.currentTimeMillis()
        while (System.currentTimeMillis() - start < timeoutMs) {
            val match = findElement(query)
            if (match.node != null) {
                return match
            }
            delay(200)
        }
        return NodeMatch(null, 0, "Timeout waiting for element", 0f)
    }

    suspend fun click(query: UIElementQuery): Pair<ActionResult, VerificationResult> {
        val startTime = System.currentTimeMillis()
        val service = AutomationAccessibilityService.getInstance()
            ?: return Pair(
                ActionResult(false, "Accessibility Service is disabled or disconnected", executionTimeMs = 0),
                VerificationResult(false, "SERVICE_DISCONNECTED", "Accessibility Service not bound")
            )

        val match = SmartNavigator.findElement(service.getActiveWindowRoot(), query)
        val targetNode = match.node
            ?: return Pair(
                ActionResult(false, "Element not found: ${match.description}", executionTimeMs = System.currentTimeMillis() - startTime),
                VerificationResult(false, "ELEMENT_NOT_FOUND", match.description)
            )

        val clicked = service.performClick(targetNode)
        delay(200) // allow UI frame transition

        val verification = if (clicked) {
            VerificationResult(true, "ACTION_CLICK_CONFIRMED: ${match.description}")
        } else {
            VerificationResult(false, "CLICK_FAILED", "Node failed to consume ACTION_CLICK")
        }

        return Pair(
            ActionResult(
                success = clicked,
                message = if (clicked) "Clicked '${targetNode.text ?: targetNode.viewIdResourceName}'" else "Click failed",
                executionTimeMs = System.currentTimeMillis() - startTime
            ),
            verification
        )
    }

    suspend fun longClick(query: UIElementQuery): Pair<ActionResult, VerificationResult> {
        val startTime = System.currentTimeMillis()
        val service = AutomationAccessibilityService.getInstance()
            ?: return Pair(
                ActionResult(false, "Accessibility Service is disconnected", executionTimeMs = 0),
                VerificationResult(false, "SERVICE_DISCONNECTED")
            )

        val match = SmartNavigator.findElement(service.getActiveWindowRoot(), query)
        val targetNode = match.node
            ?: return Pair(
                ActionResult(false, "Element not found: ${match.description}", executionTimeMs = System.currentTimeMillis() - startTime),
                VerificationResult(false, "ELEMENT_NOT_FOUND", match.description)
            )

        val longClicked = service.performLongClick(targetNode)
        delay(400)

        return Pair(
            ActionResult(
                success = longClicked,
                message = if (longClicked) "Long-clicked '${targetNode.text ?: targetNode.viewIdResourceName}'" else "Long-click failed",
                executionTimeMs = System.currentTimeMillis() - startTime
            ),
            VerificationResult(longClicked, if (longClicked) "ACTION_LONG_CLICK_CONFIRMED" else "LONG_CLICK_FAILED")
        )
    }

    suspend fun scroll(forward: Boolean = true): Pair<ActionResult, VerificationResult> {
        val startTime = System.currentTimeMillis()
        val service = AutomationAccessibilityService.getInstance()
            ?: return Pair(
                ActionResult(false, "Accessibility Service is disconnected", executionTimeMs = 0),
                VerificationResult(false, "SERVICE_DISCONNECTED")
            )

        val scrolled = service.performScroll(forward)
        return Pair(
            ActionResult(
                success = scrolled,
                message = if (scrolled) "Scrolled ${if (forward) "forward" else "backward"}" else "No scrollable container found",
                executionTimeMs = System.currentTimeMillis() - startTime
            ),
            VerificationResult(scrolled, if (scrolled) "SCROLLED" else "NOT_SCROLLABLE")
        )
    }

    suspend fun back(): Pair<ActionResult, VerificationResult> {
        val startTime = System.currentTimeMillis()
        val service = AutomationAccessibilityService.getInstance()
            ?: return Pair(
                ActionResult(false, "Accessibility Service is disconnected", executionTimeMs = 0),
                VerificationResult(false, "SERVICE_DISCONNECTED")
            )

        val backResult = service.performGlobalBack()
        return Pair(
            ActionResult(
                success = backResult,
                message = "Dispatched GLOBAL_ACTION_BACK",
                executionTimeMs = System.currentTimeMillis() - startTime
            ),
            VerificationResult(backResult, "BACK_DISPATCHED")
        )
    }
}
